### PR TITLE
Capture bailout symbols incrementally in one block

### PR DESCRIPTION
**Pull request by ThomsonTan**
Capture constant and copy prop symbols for bailout takes lots of time in regexp benchmark from Octane one some low end machine, this change caches the first capture values for bailout, and reuse it when capturing value in the following bailout in the same block.

The new bailout is constructed according to the previous bailout and changed symbol value map since then.

### CODE REVIEW

**Comment by ThomsonTan** [comment]
@LouisLaf , could you please take a look?

File: lib/Backend/GlobOpt.h [hunk 1]
```diff
@@ -957,6 +957,10 @@ class GlobOptBlockData
         argOutCount(0),
         totalOutParamCount(0),
         callSequence(nullptr),
+        capturedValuesCandicate(nullptr),
```
Comment (rajatd) [inline lib/Backend/GlobOpt.h]: typo: should be capturedValuesCandidate

File: lib/Backend/GlobOpt.h [hunk 2]
```diff
@@ -1007,6 +1011,11 @@ class GlobOptBlockData
     SListBase<IR::Opnd *> *                 callSequence;
     StackLiteralInitFldDataMap *            stackLiteralInitFldDataMap;
 
+    CapturedValues *                        capturedValuesCandicate;
```
Comment (rajatd) [inline lib/Backend/GlobOpt.h]: typo: should be capturedValuesCandidate

Comment (ThomsonTan) [inline lib/Backend/GlobOpt.h]: Right, will fix this one.

---
In reply to: [67772925](https://github.com/Microsoft/ChakraCore/pull/1152#discussion_r67772925) [](ancestors = 67772925)

File: lib/Backend/GlobOptFields.cpp [hunk 3]
```diff
@@ -2934,6 +2933,7 @@ GlobOpt::SetObjectTypeFromTypeSym(StackSym *typeSym, const Js::Type *type, Js::E
     {
         JsTypeValueInfo* valueInfo = JsTypeValueInfo::New(this->alloc, type, typeSet);
         valueInfo->SetSymStore(typeSym);
+        this->SetSymStoreDirect(valueInfo, typeSym);
```
Comment (rajatd) [inline lib/Backend/GlobOptFields.cpp]: why not replace SetSymStore with SetSymStoreDirect as you've done at other places?

Comment (ThomsonTan) [inline lib/Backend/GlobOptFields.cpp]: Good catch, the old one should be removed.

---
In reply to: [67777712](https://github.com/Microsoft/ChakraCore/pull/1152#discussion_r67777712) [](ancestors = 67777712)

File: lib/Backend/GlobOpt.cpp [hunk 4]
```diff
@@ -6935,20 +6960,34 @@ GlobOpt::SetSymStore(ValueInfo *valueInfo, Sym *sym)
...
     return sym;
 }
 
 void
+GlobOpt::SetSymStoreDirect(ValueInfo * valueInfo, Sym * sym)
+{
+    Sym * prevSymStore = valueInfo->GetSymStore();
+    if (prevSymStore && prevSymStore->IsStackSym() &&
+        prevSymStore->AsStackSym()->HasByteCodeRegSlot() &&
+        this->blockData.changedSyms)
```
Comment (satheeshravi) [inline lib/Backend/GlobOpt.cpp]: Should this be an assert instead of a pre-condition ?

Comment (ThomsonTan) [inline lib/Backend/GlobOpt.cpp]: The thought is SetSymStore doesn't depend on whether changedSyms is available or not, so it just checks here. I will consider change it to assert or not in later change.

---
In reply to: [68113107](https://github.com/Microsoft/ChakraCore/pull/1152#discussion_r68113107) [](ancestors = 68113107)

Comment (satheeshravi) [inline lib/Backend/GlobOpt.cpp]: Sure. I was suggesting to assert inside the if condition and not outside.

---
In reply to: [68113857](https://github.com/Microsoft/ChakraCore/pull/1152#discussion_r68113857) [](ancestors = 68113857,68113107)

File: lib/Backend/GlobOpt.cpp [hunk 5]
```diff
@@ -1602,6 +1609,12 @@ GlobOpt::ReuseBlockData(GlobOptBlockData *toData, GlobOptBlockData *fromData)
 
     toData->stackLiteralInitFldDataMap = fromData->stackLiteralInitFldDataMap;
 
+    // toData->capturedValues = fromData->capturedValues;
```
Comment (satheeshravi) [inline lib/Backend/GlobOpt.cpp]: nit

File: lib/Backend/GlobOpt.cpp [hunk 6]
```diff
@@ -6935,20 +6959,34 @@ GlobOpt::SetSymStore(ValueInfo *valueInfo, Sym *sym)
...
     {
-        valueInfo->SetSymStore(sym);
+        SetSymStoreDirect(valueInfo, sym);
     }
 
     return sym;
 }
 
 void
+GlobOpt::SetSymStoreDirect(ValueInfo * valueInfo, Sym * sym)
```
Comment (LouisLaf) [inline lib/Backend/GlobOpt.cpp]: Why Direct instead of just SetSymStore?  The direct part made me believe at first that this version didn't set the changed bit.

Comment (ThomsonTan) [inline lib/Backend/GlobOpt.cpp]: Direct here is no type check in the original SetSymStore, but all sym store need check changed bit. Cannot overload SetSymStore since both have identical input parameter types.

---
In reply to: [68327056](https://github.com/Microsoft/ChakraCore/pull/1152#discussion_r68327056) [](ancestors = 68327056)

File: lib/Backend/GlobOpt.cpp [hunk 7]
```diff
@@ -1060,6 +1060,9 @@ GlobOpt::MergePredBlocksValueMaps(BasicBlock *block)
         this->KillAllObjectTypes();
     }
 
+    this->blockData.capturedArgs = JitAnew(this->alloc, BVSparse<JitArenaAllocator>, this->alloc);
+    this->blockData.changedSyms = JitAnew(this->alloc, BVSparse<JitArenaAllocator>, this->alloc);
```
Comment (LouisLaf) [inline lib/Backend/GlobOpt.cpp]: This should go in InitBlockData().  The Clone and Reuse path will have valid BVs and you are allocating new ones unnecessarily.

File: lib/Backend/GlobOpt.cpp (hunk 6 above)
Comment (LouisLaf) [inline lib/Backend/GlobOpt.cpp]: Oh I get it.  I miss the fact there already was a SetSymStore on GLobOpt.

---
In reply to: [68327517](https://github.com/Microsoft/ChakraCore/pull/1152#discussion_r68327517) [](ancestors = 68327517,68327056)

File: lib/Backend/GlobOpt.cpp (hunk 4 above)
Comment (LouisLaf) [inline lib/Backend/GlobOpt.cpp]: I don't see a path where changedSyms would be NULL.  You should remove the NULL checks.

---
In reply to: [68132431](https://github.com/Microsoft/ChakraCore/pull/1152#discussion_r68132431) [](ancestors = 68132431,68113857,68113107)

File: lib/Backend/GlobOpt.cpp [hunk 8]
```diff
@@ -6957,6 +6995,10 @@ GlobOpt::SetValue(GlobOptBlockData *blockData, Value *val, Sym * sym)
     else
     {
         SetValueToHashTable(blockData->symToValueMap, val, sym);
+        if (blockData->changedSyms && isStackSym && sym->AsStackSym()->HasByteCodeRegSlot())
```
Comment (LouisLaf) [inline lib/Backend/GlobOpt.cpp]: Here too.

File: lib/Backend/GlobOpt.cpp (hunk 7 above)
Comment (ThomsonTan) [inline lib/Backend/GlobOpt.cpp]: Yes, reuse is good. Will take this change in the cross block version.

---
In reply to: [68328456](https://github.com/Microsoft/ChakraCore/pull/1152#discussion_r68328456) [](ancestors = 68328456)

File: lib/Backend/GlobOpt.cpp (hunk 4 above)
Comment (ThomsonTan) [inline lib/Backend/GlobOpt.cpp]: Will try remove it when propagating values across blocks, it should be fine if its construction is moved to InitBlockData.

---
In reply to: [68328835](https://github.com/Microsoft/ChakraCore/pull/1152#discussion_r68328835) [](ancestors = 68328835,68132431,68113857,68113107)

File: lib/Backend/GlobOpt.cpp (hunk 8 above)
Comment (ThomsonTan) [inline lib/Backend/GlobOpt.cpp]: Will do it.

---
In reply to: [68328865](https://github.com/Microsoft/ChakraCore/pull/1152#discussion_r68328865) [](ancestors = 68328865)

File: lib/Backend/GlobOptBailOut.cpp [hunk 9]
```diff
@@ -15,52 +15,258 @@ GlobOpt::CaptureValue(BasicBlock *block, StackSym * stackSym, Value * value, Bai
...
+            Sym * sym = hasConstValue ? iterConst.Data().Key() : nullptr;
+            Value * val = nullptr;
+            HashBucket<Sym *, Value *> * symIdBucket = nullptr;
+            bool symIdHandled = false;
+
+            // copy unchanged sym to new capturedValues
+            while (sym && sym->m_id < symId)
+            {
+                Assert(sym->IsStackSym());
+                if (!sym->AsStackSym()->HasArgSlotNum() || block->globOptData.capturedArgs->Test(sym->m_id))
```
Comment (LouisLaf) [inline lib/Backend/GlobOptBailOut.cpp]: This should always be false, otherwise the sym would be in the changedSym bit vector.

File: lib/Backend/GlobOptBailOut.cpp [hunk 10]
```diff
@@ -15,52 +15,258 @@ GlobOpt::CaptureValue(BasicBlock *block, StackSym * stackSym, Value * value, Bai
...
+                    continue;
+                }
+
+                sym = symIdBucket->value;
+                Assert(sym->IsStackSym() && (sym->AsStackSym()->HasByteCodeRegSlot() || sym->AsStackSym()->HasArgSlotNum()));
+
+                val =  block->globOptData.capturedArgs->Test(symId) ? FindValue(sym) : symIdBucket->element;
+                ValueInfo* valueInfo = val->GetValueInfo();
+
+                if (valueInfo->GetSymStore() != nullptr)
```
Comment (LouisLaf) [inline lib/Backend/GlobOptBailOut.cpp]: Is this necessary?

File: lib/Backend/GlobOptBailOut.cpp [hunk 11]
```diff
@@ -15,52 +15,258 @@ GlobOpt::CaptureValue(BasicBlock *block, StackSym * stackSym, Value * value, Bai
...
+                {
+                    int32 intConstValue;
+                    BailoutConstantValue constValue;
+
+                    if (valueInfo->TryGetIntConstantValue(&intConstValue))
+                    {
+                        constValue.InitIntConstValue(intConstValue);
+                        bailOutConstValuesIter.InsertNodeBefore(this->func->m_alloc, sym->AsStackSym(), constValue);
+                        symIdHandled = true;
```
Comment (LouisLaf) [inline lib/Backend/GlobOptBailOut.cpp]: Get rid of symIdHandled and just put continue here.

File: lib/Backend/GlobOptBailOut.cpp [hunk 12]
```diff
@@ -15,52 +15,258 @@ GlobOpt::CaptureValue(BasicBlock *block, StackSym * stackSym, Value * value, Bai
...
+                StackSym * copyPropSym = (sym != nullptr) ? iterCopyPropSym.Data().Value() : nullptr;
+                Sym * mergeSym = nullptr;
+                bool mergeSymId = false;
+
+                // copy unchanged copy prop sym
+                if (sym && !tempBv->Test(copyPropSym->m_id) && sym->m_id < symId)
+                {
+                    Assert(sym->IsStackSym());
+
+                    if (!sym->AsStackSym()->HasArgSlotNum() || block->globOptData.capturedArgs->Test(sym->m_id))
```
Comment (LouisLaf) [inline lib/Backend/GlobOptBailOut.cpp]: Again this should always be false.

File: lib/Backend/GlobOptBailOut.cpp [hunk 13]
```diff
@@ -15,52 +15,258 @@ GlobOpt::CaptureValue(BasicBlock *block, StackSym * stackSym, Value * value, Bai
...
+            sym = hasCopyPropSym ? iterCopyPropSym.Data().Key() : nullptr;
+
+            while (true)
+            {
+                StackSym * copyPropSym = (sym != nullptr) ? iterCopyPropSym.Data().Value() : nullptr;
+                Sym * mergeSym = nullptr;
+                bool mergeSymId = false;
+
+                // copy unchanged copy prop sym
+                if (sym && !tempBv->Test(copyPropSym->m_id) && sym->m_id < symId)
```
Comment (LouisLaf) [inline lib/Backend/GlobOptBailOut.cpp]: Put the Bv test at the end to avoid it when sym->m_id is >= to symId.

File: lib/Backend/GlobOptBailOut.cpp [hunk 14]
```diff
@@ -15,52 +15,258 @@ GlobOpt::CaptureValue(BasicBlock *block, StackSym * stackSym, Value * value, Bai
...
+            }
+
+            if (symIdHandled)
+            {
+                continue;
+            }
+
+            sym = hasCopyPropSym ? iterCopyPropSym.Data().Key() : nullptr;
+
+            while (true)
```
Comment (LouisLaf) [inline lib/Backend/GlobOptBailOut.cpp]: This code is really hard to follow.  Why didn't you lay it out like the constant case above:
1. loop over the unchanged syms
2. captureValue() on the symId
3. skip to next if sym->m_id == symId

That while(true) should only loop over the unchanged syms.  Put the rest of the code out of that loop.

File: lib/Backend/GlobOptFields.cpp [hunk 15]
```diff
@@ -689,8 +689,7 @@ GlobOpt::PreparePrepassFieldHoisting(Loop * loop)
 
                 // Set object type live on prepass so we can track if it got killed in the loop. (see FinishOptHoistedPropOps)
                 JsTypeValueInfo* typeValueInfo = JsTypeValueInfo::New(this->alloc, nullptr, nullptr);
-                typeValueInfo->SetSymStore(typeSym);
-                typeValueInfo->SetIsShared();
```
Comment (LouisLaf) [inline lib/Backend/GlobOptFields.cpp]: Why did you get rid of SetIsShared()?

File: lib/Backend/GlobOptBailOut.cpp (hunk 9 above)
Comment (ThomsonTan) [inline lib/Backend/GlobOptBailOut.cpp]: Yes, this is useless code.

---
In reply to: [68345673](https://github.com/Microsoft/ChakraCore/pull/1152#discussion_r68345673) [](ancestors = 68345673)

File: lib/Backend/GlobOptBailOut.cpp (hunk 10 above)
Comment (ThomsonTan) [inline lib/Backend/GlobOptBailOut.cpp]: This check is inherited from the previous version.

---
In reply to: [68345902](https://github.com/Microsoft/ChakraCore/pull/1152#discussion_r68345902) [](ancestors = 68345902)

File: lib/Backend/GlobOptBailOut.cpp (hunk 11 above)
Comment (ThomsonTan) [inline lib/Backend/GlobOptBailOut.cpp]: Will remove it.

---
In reply to: [68346060](https://github.com/Microsoft/ChakraCore/pull/1152#discussion_r68346060) [](ancestors = 68346060)

File: lib/Backend/GlobOptBailOut.cpp (hunk 13 above)
Comment (ThomsonTan) [inline lib/Backend/GlobOptBailOut.cpp]: Will update the order, my previous thought is put condition with more selectivity (not almost always true) in front.

---
In reply to: [68346499](https://github.com/Microsoft/ChakraCore/pull/1152#discussion_r68346499) [](ancestors = 68346499)

Comment (LouisLaf) [inline lib/Backend/GlobOptBailOut.cpp]: The Bv test is much more expensive.

---
In reply to: [68348778](https://github.com/Microsoft/ChakraCore/pull/1152#discussion_r68348778) [](ancestors = 68348778,68346499)

File: lib/Backend/GlobOptBailOut.cpp (hunk 12 above)
Comment (ThomsonTan) [inline lib/Backend/GlobOptBailOut.cpp]: Will remove this.

---
In reply to: [68346272](https://github.com/Microsoft/ChakraCore/pull/1152#discussion_r68346272) [](ancestors = 68346272)

File: lib/Backend/GlobOptBailOut.cpp (hunk 14 above)
Comment (ThomsonTan) [inline lib/Backend/GlobOptBailOut.cpp]: The structure is similar, but this loop need to handle more things because the change in copy sym will also trigger recapture path before reaching changed symId.

---
In reply to: [68347583](https://github.com/Microsoft/ChakraCore/pull/1152#discussion_r68347583) [](ancestors = 68347583)

File: lib/Backend/GlobOptFields.cpp (hunk 15 above)
Comment (ThomsonTan) [inline lib/Backend/GlobOptFields.cpp]: Oh, this is not intended, will add it back. Thanks for the catch.

---
In reply to: [68347957](https://github.com/Microsoft/ChakraCore/pull/1152#discussion_r68347957) [](ancestors = 68347957)

File: lib/Backend/GlobOptBailOut.cpp (hunk 14 above)
Comment (LouisLaf) [inline lib/Backend/GlobOptBailOut.cpp]: If copy-sym changed but the sym itself didn't, then all you have to do is not add sym to the CopySyms list in the loop copying the unchanged syms.  

---
In reply to: [68350810](https://github.com/Microsoft/ChakraCore/pull/1152#discussion_r68350810) [](ancestors = 68350810,68347583)

File: lib/Backend/GlobOpt.cpp (hunk 7 above)
Comment (ThomsonTan) [inline lib/Backend/GlobOpt.cpp]: InitDataBlockData is called when there is no predecessor. Still need to create new BV when have predecessor so InitDataBlockData doesn't meet the requirement.

---
In reply to: [68328885](https://github.com/Microsoft/ChakraCore/pull/1152#discussion_r68328885) [](ancestors = 68328885,68328456)

Comment (LouisLaf) [inline lib/Backend/GlobOpt.cpp]: Oh, I though you already had code in CloneBlockData but it is in CopyBlockData.
What I mean is that you should put this in InitBlockData/CloneBlockData/ReuseBlockData like all the other data structures.  This will allow better reuse.

---
In reply to: [68686479](https://github.com/Microsoft/ChakraCore/pull/1152#discussion_r68686479) [](ancestors = 68686479,68328885,68328456)

File: lib/Backend/GlobOptBailOut.cpp [hunk 16]
```diff
@@ -5,62 +5,247 @@
...
+
+    CapturedValues capturedValues;
+    SListBase<ConstantStackSymValue>::EditingIterator bailOutConstValuesIter(&capturedValues.constantValues);
+    SListBase<CopyPropSyms>::EditingIterator bailOutCopySymsIter(&capturedValues.copyPropSyms);
+
+    bailOutConstValuesIter.Next();
+    bailOutCopySymsIter.Next();
+
+    // tempBv has both changed syms with bytecode reg slot or arg slot
+    BVSparse<JitArenaAllocator> * tempBv = JitAnew(this->tempAlloc, BVSparse<JitArenaAllocator>, this->tempAlloc);
```
Comment (LouisLaf) [inline lib/Backend/GlobOptBailOut.cpp]: Rename to something like changedSymsAndArgs.

File: lib/Backend/GlobOptBailOut.cpp [hunk 17]
```diff
@@ -5,62 +5,247 @@
...
+
+    if (!block->globOptData.capturedValues)
     {
```
Comment (LouisLaf) [inline lib/Backend/GlobOptBailOut.cpp]: Factor the code into:
if ()
{
    CaptureValueFull()
} else {
   foreach_bits() {
        ProcessCaptureVarConstValue()
        ProcessCaptureCopoySymValue()
    }
}

File: lib/Backend/GlobOptBailOut.cpp [hunk 18]
```diff
@@ -5,62 +5,247 @@
...
+                    }
+                }
+
+                hasCopyPropSym = iterCopyPropSym.Next();
+                sym = hasCopyPropSym ? iterCopyPropSym.Data().Key() : nullptr;
+            }
+            if (sym && sym->m_id == symId)
+            {
+                hasCopyPropSym = iterCopyPropSym.Next();
+            }
```
Comment (LouisLaf) [inline lib/Backend/GlobOptBailOut.cpp]: Moved to end of loop as it prepares the next iteration.

File: lib/Backend/GlobOptBailOut.cpp [hunk 19]
```diff
@@ -5,62 +5,247 @@
...
+                }
+
+                hasConstValue = iterConst.Next();
+                sym = hasConstValue ? iterConst.Data().Key() : nullptr;
+            }
+            if (sym && sym->m_id == symId)
+            {
+                hasConstValue = iterConst.Next();
+            }
+            if (symId != Js::Constants::InvalidSymID)
```
Comment (LouisLaf) [inline lib/Backend/GlobOptBailOut.cpp]: Moved to end of loop as it prepares the next iteration.

File: lib/Backend/GlobOptBailOut.cpp [hunk 20]
```diff
@@ -5,62 +5,247 @@
...
+            {
+                hasCopyPropSym = iterCopyPropSym.Next();
+            }
+            if (symId != Js::Constants::InvalidSymID)
+            {
+                // recapture changed copy prop sym
+                symIdBucket = block->globOptData.symToValueMap->GetBucket(symId);
+                if (symIdBucket != nullptr)
+                {
+                    sym = symIdBucket->value;
```
Comment (LouisLaf) [inline lib/Backend/GlobOptBailOut.cpp]: Rename as this 'sym' is not at all the same 'sym' as before.

**Comment by LouisLaf** [comment]
:shipit:
